### PR TITLE
Fix #775: Cannot find reports with relative Ant globbing

### LINKED ISSUE #775: Cannot find reports with relative Ant globbing
Environment:
- SQ 5.3
- 2ba787e23cb791a414391150fa269b45fa38e454 (latest from master at time of writing)
- Multi-module project

This multi-module project is analyzed using the SQ Gradle Scanner.  When all said and done, the settings end up looking something like this:

```
sonar.modules=module1,module2
...
module2.sonar.sources=dir1,dir2,dir3,dir4
module2.sonar.exclusions=**/build/**,**/cppcheck.xml
module2.sonar.cxx.cppcheck.reportPath=**/cppcheck.xml
```

The reports most certainly exist in the tree (and they have correct ownership and 0644 perms):

```
$ find . -name cppcheck.xml
./module2/dir1/build/cppcheck.xml
./module2/dir2/cppcheck.xml
./module2/dir3/build/cppcheck.xml
./module2/dir4/build/cppcheck.xml
```

But here's what I get in the analysis:

```
-------------  Scan module2
Base dir: /path/to/project/module2
Working dir: /path/to/project/build/sonar/com.mycompany.project_module2
Source paths: dir1, dir2, dir3, dir4
Source encoding: ISO-8859-1, default locale: en_US
Index files
Excluded sources: 
  */build/**
  **/cppcheck.xml
82 files indexed
...
omitted for brevity
...
Sensor CxxCppCheckSensor
Using pattern '**/cppcheck.xml' to find reports
Unprocessed root directory '/path/to/project/module2'
Unprocessed report file '**/cppcheck.xml'
Processed root directory '/path/to/project/module2'
Processed report file '**/cppcheck.xml'
Cannot find a report for 'sonar.cxx.cppcheck.reportPath=**/cppcheck.xml'
Sensor CxxCppCheckSensor (done) | time=29ms
```

As you can see, the property makes its way to the sensor correctly but something goes wrong when actually searching for the files.  There are no exceptions when walking the tree.  I don't know exactly where the problem in the code is without attaching a debugger, but it's within this stack: https://github.com/SonarOpenCommunity/sonar-cxx/blob/2ba787e23cb791a414391150fa269b45fa38e454/sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxUtils.java#L119 with ```CxxFileFinder.FindFiles``` turning up an empty Collection.

### PR DESCRIPTION
**Pull request by mjdetullio**
Upgrade Ant dependency to latest.  Newer versions of DirectoryScanner do
not require a basedir to be set if all includes are absolute, which
makes this fix much easier.

Normalize the report paths and make them all absolute.  Relative
reports/globs are made absolute using the module's base dir.

Use Ant's DirectoryScanner with the now formatted report paths and no
basedir to allow globbing/wildcards on any location on the filesystem.

### CODE REVIEW

**Comment by guwirth** [comment]
@jmecosta do you know what the root cause for this incompatibility is?

**Comment by jmecosta** [comment]
Likely ant dependency update to latest? Or maybe it's just need to be checked and fixed in this pull request?

**Comment by mjdetullio** [comment]
Small bug... the `moduleBaseDir` needed to be normalized as well.  Caused problems when the analysis directory used `.` or `../` backticks.  The ITs were providing module dirs like `/path/to/sonar-cxx/integration-tests/testdata/googletest_project/./`

With that out of the way...

The bigger problem is some of the ITs need updated.  For example, some depend on a logging statement that no longer exists.  `Cannot find a report for <entry>` isn't reported anymore, and isn't possible with this implementation since it scans for all entries at once, rather than scanning for each individual report entry.  (An entry in this context being one of the values in the comma-separated list of reportPaths)

```
$ grep -R "Cannot find a report for" integration-tests/
integration-tests//features/importing_coverage.feature:              .*WARN.*Cannot find a report for '.*'
integration-tests//features/multimodule_analysis.feature:              .*WARN.*Cannot find a report for '.*'
integration-tests//features/multimodule_analysis_2.feature:#              .*WARN.*Cannot find a report for '.*'
integration-tests//features/smoketest.feature:              .*WARN.*Cannot find a report for '.*'
integration-tests//features/test_create_rules_and_bullseye_import.feature:              .*WARN.*Cannot find a report for '.*'
integration-tests//features/test_execution_statistics.feature:              .*WARN.*Cannot find a report for '.*'
```

**Comment by jmecosta** [comment]
Cannot find a report for <entry> => are allowed warning messages during a test, if they dont show up i dont see any problem with it.

the its look ok to me now

**Comment by mjdetullio** [comment]
I just noticed that as well and I removed the log line exceptions from those tests.

**Comment by jmecosta** [comment]
looks good to me now.

File: sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java [hunk 1]
```diff
@@ -142,25 +147,36 @@ protected String getStringProperty(String name, String def) {
...
-      for (String reportPath : reportPaths) {
-        reportPath = FilenameUtils.normalize(reportPath);
-        File singleFile = new File(reportPath);
-        if (singleFile.exists()) {
-          reports.add(singleFile);
-        } else {
-          CxxUtils.LOG.debug("Using pattern '{}' to find reports", reportPath);
-          CxxUtils.GetReportForBaseDirAndPattern(moduleBaseDir.getPath(), reportPath, reports);
-          if (reports.isEmpty()) {
-            CxxUtils.LOG.warn("Cannot find a report for '{}={}'", reportPathPropertyKey, reportPath);
```
Comment (guwirth) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: this warning is missing in new solution

File: sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java [hunk 2]
```diff
@@ -142,25 +147,36 @@ protected String getStringProperty(String name, String def) {
...
+      CxxUtils.LOG.debug("Normalized report includes: {}", includes);
+
+      DirectoryScanner directoryScanner = new DirectoryScanner();
+      directoryScanner.setIncludes(includes.toArray(new String[includes.size()]));
+      directoryScanner.scan();
+
+      for (String found : directoryScanner.getIncludedFiles()) {
+        CxxUtils.LOG.debug("Adding report {}", found);
+        reports.add(new File(found));
       }
```
Comment (guwirth) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: why not adding the warning again when `reports` is empty after for loop?

e.g.
```Java
CxxUtils.LOG.warn("Cannot find a report for '{}'", reportPathPropertyKey);
```

Comment (mjdetullio) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: Good suggestion

**Comment by mjdetullio** [comment]
Updated based on feedback from @guwirth

**Comment by guwirth** [comment]
:+1: will merge it this evening.
If others like to review do it till this evening please.
Fix for #775

**Comment by guwirth** [comment]
@mjdetullio thanks a lot for your contribution.

**Comment by mjdetullio** [comment]
You're welcome :+1: 

Thanks for maintaining :)

**Pull request merged**